pmd-core/src/main/java/net/sourceforge/pmd/PMD.java

```diff
@@ -284,73 +221,148 @@ private static void printRuleNamesInDebug(List<RuleSet> rulesets) {
     }
 
     /**
-     * Creates a new rule context, initialized with a new, empty report.
+     * Run PMD using the given configuration. This replaces the other overload.
+     *
+     * @param configuration Configuration for the run. Note that the files,
+     *                      and rulesets, are ignored, as they are supplied
+     *                      as parameters
+     * @param ruleSets      Parsed rulesets
+     * @param files         Files to process, will be closed by this method.
+     * @param renderers     Renderers that render the report (may be empty)
      *
-     * @param sourceCodeFilename
-     *            the source code filename
-     * @param sourceCodeFile
-     *            the source code file
-     * @return the rule context
+     * @return Report in which violations are accumulated
      *
-     * @deprecated Not useful
+     * @throws Exception If there was a problem when opening or closing the renderers
      */
-    @Deprecated
-    public static RuleContext newRuleContext(String sourceCodeFilename, File sourceCodeFile) {
+    @SuppressWarnings("PMD.CloseResource")
+    public static Report processFiles(PMDConfiguration configuration,
+                                      List<RuleSet> ruleSets,
+                                      List<DataSource> files,
```

Comment:
I was thinking maybe it would be sensible to have a special class that encapsulates the set of data sources/text documents. I think this would be needed in order to encapsulate callbacks to close resources from which the text documents are taken (eg a zip file). That way even if something fails in the analysis they can still be closed in a finally block.

If we could introduce this on master, it could encapsulate the file collection phase, and provide a forward-compatible API. For instance:
```java
interface FileCollector implements AutoCloseable {

    // add a file, language version is determined automatically from the extension/file patterns
    void addFileOrDirectory(Path file);
   // add a file and force the given language to override the file extensions
    void addFile(Path file, LanguageVersion language);

    // opens the zip file and returns a FileSystem for its contents, so you can explore it 
    // with the nio.Path API and call FileCollector::addFile. The FileCollector closes the
    // FileSystem instance in its own close method.
    FileSystem addZipFile(Path zipFile);

    // override the default file extensions for a language
    // may be a programmatic solution to bypass file extensions, eg pom.xml  
    void setFilePatterns(Language language, Pattern pattern);

    // close registered resources (eg zip files)
    void close();
}
```
Internally this will build a set of DataSources in PMD 6, which we can change in PMD 7

Reply:
Additional note: it would be nice to design this API so that CPD can also use it.

Reply:
:+1: Sounds like a good idea. This should make the class "PMD" less crowded.
At some point, we need to draw a line between these two "APIs":
* The API to programmatically execute PMD - here I assume, you already pass in DataSources or whatever. This is (as it is now) in pmd-core
* The CLI, which calls the API above. Here you pass in a filesystem path to the sources and this part does the file collection. Or in case you pass in a file list, it would do the conversion into DataSources. This would be in an own module (together with the ant task). To have this separated (in PMD 7) has some benefit for slf4j/logging impl - as we don't want to add a transitive dependency for a logging implementation to a library (pmd-core is essentially a library).

Reply:
I think, we need to keep the parameter type to be `Collection<? extends DataSource>`. In the build, we now get a `java.lang.NoSuchMethodError: 'net.sourceforge.pmd.Report net.sourceforge.pmd.PMD.processFiles(net.sourceforge.pmd.PMDConfiguration, java.util.List, java.util.Collection, java.util.List)'` when we try to run maven-pmd-plugin with PMD7.
https://github.com/pmd/pmd/runs/5231261489?check_suite_focus=true#step:7:11149

I'll give it a try, whether it works when I restore this parameter.

We can maybe deprecate this method when we have the FileCollector API in.